### ISSUE TITLE
7.4.2299 causes lgetexpr/laddexpr to fail after a while (test suite)

### ISSUE THREAD

**Issue by blueyed** [vim/vim#1225]
I've noticed that 7.4.2299 (04c4ce6) causes the Neomake tests to fail out of a sudden.

It is related to using `laddexpr ''` etc and shows itself when `laddexpr lines` suddenly stops working: `getloclist(0)` returns the old list then still.

The setup for this is rather complex, but can be reproduced using `make testvim` (on this PR: https://github.com/neomake/neomake/pull/751).
It gets triggered by one of the last tests, and moving that one to the beginning (or commenting a bunch of others before it) works around this.

The following patch fixes it:
```diff
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git i/src/quickfix.c w/src/quickfix.c
index 2f5256b..853e621 100644
--- i/src/quickfix.c
+++ w/src/quickfix.c
@@ -4964,7 +4964,7 @@ ex_cexpr(exarg_T *eap)
        apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
                                               curbuf->b_fname, TRUE, curbuf);
 # ifdef FEAT_EVAL
-       if (did_throw || force_abort)
+       if (did_throw)
            return;
 # endif
     }
```

There are more usages of `force_abort` in there, and this maybe only shadows another issue?!

How would I debug it from there?

The `if (did_throw || force_abort) return;` pattern is not used often, but only in quickfix.c and `src/if_cscope.c`: https://github.com/vim/vim/blob/04c4ce650f9e533cd35b2aa6803f4d354d3ec7aa/src/if_cscope.c#L1190-L1191.
Maybe it's just wrong like that?

/cc @yegappan (as the author of the patch?!)

**Comment by blueyed** [comment]
This would also help:

``` diff
diff --git i/src/quickfix.c w/src/quickfix.c
index 2f5256b..91b05c4 100644
--- i/src/quickfix.c
+++ w/src/quickfix.c
@@ -4961,10 +4961,10 @@ ex_cexpr(exarg_T *eap)
     }
     if (au_name != NULL)
     {
-       apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
+       int ret = apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
                                               curbuf->b_fname, TRUE, curbuf);
 # ifdef FEAT_EVAL
-       if (did_throw || force_abort)
+       if (ret && (did_throw || force_abort))
            return;
 # endif
     }
```

**Comment by blueyed** [comment]
Here is a test case:

``` vim
set errorformat=%m
lgetexpr '?'

try
  call DoesNotExit()
catch
  lgetexpr '1'
finally
  lgetexpr '1'
endtry

if getloclist(0)[0].text != 1
  echoerr 'laddexpr was not called'
endif
```

What about using `aborting()` instead of `did_throw || force_abort`?
But even then it seems strange to handle it there like this, especially for cases where no autocommand has been called (and the status might not have been checked before, in case some ran successfully).

**Comment by vim-ml** [comment]
Hi Daniel,

On Fri, Nov 4, 2016 at 1:50 PM, Daniel Hahler vim-dev-github@256bit.org wrote:

> Here is a test case:
> 
> set errorformat=%m
> lgetexpr '?'
> 
> try
>   call DoesNotExit()
> catch
>   lgetexpr '1'
> finally
>   lgetexpr '1'
> endtry
> 
> if getloclist(0)[0].text != 1
>   echoerr 'laddexpr was not called'
> endif
> 
> What about using aborting() instead of did_throw || force_abort?

It looks like using aborting() in this case fixes the issue. All the quickfix
related tests are also passing. But I am not sure about the side effects
of this change.

> But even then it seems strange to handle it there like this, especially for
> cases where no autocommand has been called (and the status might not have
> been checked before, in case some ran successfully).

I think the code can be changed to the following:

   if (apply_autocmds(....))
              if (aborting())
                    return;
- Yegappan

**Comment by brammool** [comment]
Daniel Hahler wrote:

> Here is a test case:
> 
> ``` vim
> set errorformat=%m
> lgetexpr '?'
> 
> try
>   call DoesNotExit()
> catch
>   lgetexpr '1'
> finally
>   lgetexpr '1'
> endtry
> 
> if getloclist(0)[0].text != 1
>   echoerr 'laddexpr was not called'
> endif
> ```
> 
> What about using `aborting()` instead of `did_throw || force_abort`?
> But even then it seems strange to handle it there like this,
> especially for cases where no autocommand has been called (and the
> status might not have been checked before, in case some ran
> successfully).

We can check if apply_autocmds returns TRUE and use aborting().
How about this:

**\* ../vim-8.0.0065/src/quickfix.c  2016-10-15 19:28:09.239181100 +0200
--- src/quickfix.c  2016-11-05 20:29:05.956598800 +0100

---

**\* 3446,3457 ****
    case CMD_lgrepadd:  au_name = (char_u *)"lgrepadd"; break;
    default: break;
      }
!     if (au_name != NULL)
      {
-   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf);
  # ifdef FEAT_EVAL
  !   if (did_throw || force_abort)
      return;
  # endif
      }
  --- 3446,3456 ----
  case CMD_lgrepadd:  au_name = (char_u _)"lgrepadd"; break;
  default: break;
    }
  !     if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
  !                          curbuf->b_fname, TRUE, curbuf))
    {
  # ifdef FEAT_EVAL
  !   if (aborting())
      return;
  # endif
      }
  *_*************
  **\* 3970,3981 ****
  case CMD_lgrepadd:    au_name = (char_u *)"lgrepadd"; break;
  default: break;
    }
  !     if (au_name != NULL)
    {
  !   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
  !                          curbuf->b_fname, TRUE, curbuf);
  !   if (did_throw || force_abort)
      return;
    }
  #endif

--- 3969,3981 ----
    case CMD_lgrepadd:    au_name = (char_u *)"lgrepadd"; break;
    default: break;
      }
!     if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
!                          curbuf->b_fname, TRUE, curbuf))
      {
! # ifdef FEAT_EVAL
!   if (aborting())
        return;
- # endif
    }
  #endif

---

**\* 4875,4886 ****
    case CMD_laddbuffer:    au_name = (char_u *)"laddbuffer"; break;
    default: break;
      }
!     if (au_name != NULL)
      {
-   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf);
  # ifdef FEAT_EVAL
  !   if (did_throw || force_abort)
      return;
  # endif
      }
  --- 4875,4885 ----
  case CMD_laddbuffer:    au_name = (char_u _)"laddbuffer"; break;
  default: break;
    }
  !     if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
  !                          curbuf->b_fname, TRUE, curbuf))
    {
  # ifdef FEAT_EVAL
  !   if (aborting())
      return;
  # endif
      }
  *_*************
  **\* 4966,4977 ****
  case CMD_laddexpr:  au_name = (char_u *)"laddexpr"; break;
  default: break;
    }
  !     if (au_name != NULL)
    {
-   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf);
  # ifdef FEAT_EVAL
  !   if (did_throw || force_abort)
      return;
  # endif
      }
  --- 4965,4975 ----
  case CMD_laddexpr:  au_name = (char_u _)"laddexpr"; break;
  default: break;
    }
  !     if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
  !                          curbuf->b_fname, TRUE, curbuf))
    {
  # ifdef FEAT_EVAL
  !   if (aborting())
      return;
  # endif
      }
  *_*************
  **\* 5042,5053 ****
  case CMD_lhelpgrep: au_name = (char_u *)"lhelpgrep"; break;
  default: break;
    }
  !     if (au_name != NULL)
    {
  !   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
  !                          curbuf->b_fname, TRUE, curbuf);
  !   if (did_throw || force_abort)
      return;
    }
  #endif

--- 5040,5052 ----
    case CMD_lhelpgrep: au_name = (char_u *)"lhelpgrep"; break;
    default: break;
      }
!     if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
!                          curbuf->b_fname, TRUE, curbuf))
      {
! # ifdef FEAT_EVAL
!   if (aborting())
        return;
- # endif
    }
  #endif

**\* ../vim-8.0.0065/src/quickfix.c  2016-10-15 19:28:09.239181100 +0200
--- src/quickfix.c  2016-11-05 20:29:05.956598800 +0100

---

**\* 3446,3457 ****
    case CMD_lgrepadd:  au_name = (char_u *)"lgrepadd"; break;
    default: break;
      }
!     if (au_name != NULL)
      {
-   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf);
  # ifdef FEAT_EVAL
  !   if (did_throw || force_abort)
      return;
  # endif
      }
  --- 3446,3456 ----
  case CMD_lgrepadd:  au_name = (char_u _)"lgrepadd"; break;
  default: break;
    }
  !     if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
  !                          curbuf->b_fname, TRUE, curbuf))
    {
  # ifdef FEAT_EVAL
  !   if (aborting())
      return;
  # endif
      }
  *_*************
  **\* 3970,3981 ****
  case CMD_lgrepadd:    au_name = (char_u *)"lgrepadd"; break;
  default: break;
    }
  !     if (au_name != NULL)
    {
  !   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
  !                          curbuf->b_fname, TRUE, curbuf);
  !   if (did_throw || force_abort)
      return;
    }
  #endif

--- 3969,3981 ----
    case CMD_lgrepadd:    au_name = (char_u *)"lgrepadd"; break;
    default: break;
      }
!     if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
!                          curbuf->b_fname, TRUE, curbuf))
      {
! # ifdef FEAT_EVAL
!   if (aborting())
        return;
- # endif
    }
  #endif

---

**\* 4875,4886 ****
    case CMD_laddbuffer:    au_name = (char_u *)"laddbuffer"; break;
    default: break;
      }
!     if (au_name != NULL)
      {
-   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf);
  # ifdef FEAT_EVAL
  !   if (did_throw || force_abort)
      return;
  # endif
      }
  --- 4875,4885 ----
  case CMD_laddbuffer:    au_name = (char_u _)"laddbuffer"; break;
  default: break;
    }
  !     if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
  !                          curbuf->b_fname, TRUE, curbuf))
    {
  # ifdef FEAT_EVAL
  !   if (aborting())
      return;
  # endif
      }
  *_*************
  **\* 4966,4977 ****
  case CMD_laddexpr:  au_name = (char_u *)"laddexpr"; break;
  default: break;
    }
  !     if (au_name != NULL)
    {
-   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf);
  # ifdef FEAT_EVAL
  !   if (did_throw || force_abort)
      return;
  # endif
      }
  --- 4965,4975 ----
  case CMD_laddexpr:  au_name = (char_u _)"laddexpr"; break;
  default: break;
    }
  !     if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
  !                          curbuf->b_fname, TRUE, curbuf))
    {
  # ifdef FEAT_EVAL
  !   if (aborting())
      return;
  # endif
      }
  *_*************
  **\* 5042,5053 ****
  case CMD_lhelpgrep: au_name = (char_u *)"lhelpgrep"; break;
  default: break;
    }
  !     if (au_name != NULL)
    {
  !   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
  !                          curbuf->b_fname, TRUE, curbuf);
  !   if (did_throw || force_abort)
      return;
    }
  #endif

--- 5040,5052 ----
    case CMD_lhelpgrep: au_name = (char_u *)"lhelpgrep"; break;
    default: break;
      }
!     if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
!                          curbuf->b_fname, TRUE, curbuf))
      {
! # ifdef FEAT_EVAL
!   if (aborting())
        return;
- # endif
    }
  #endif

## 

Engineers are always delighted to share wisdom, even in areas in which they
have no experience whatsoever.  Their logic provides them with inherent
insight into any field of expertise.  This can be a problem when dealing with
the illogical people who believe that knowledge can only be derived through
experience.
                (Scott Adams - The Dilbert principle)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by vim-ml** [comment]
Hi Bram,

On Sat, Nov 5, 2016 at 12:59 PM, Bram Moolenaar
vim-dev-github@256bit.org wrote:

> Daniel Hahler wrote:
> 
> > Here is a test case:
> > 
> > ``` vim
> > set errorformat=%m
> > lgetexpr '?'
> > 
> > try
> > call DoesNotExit()
> > catch
> > lgetexpr '1'
> > finally
> > lgetexpr '1'
> > endtry
> > 
> > if getloclist(0)[0].text != 1
> > echoerr 'laddexpr was not called'
> > endif
> > ```
> > 
> > What about using `aborting()` instead of `did_throw || force_abort`?
> > But even then it seems strange to handle it there like this,
> > especially for cases where no autocommand has been called (and the
> > status might not have been checked before, in case some ran
> > successfully).
> 
> We can check if apply_autocmds returns TRUE and use aborting().
> How about this:

Your change looks good to me. I am attaching a diff with your changes
and the test from Daniel.
- Yegappan

diff --git a/src/quickfix.c b/src/quickfix.c
index 5bd1257..b303b54 100644
--- a/src/quickfix.c
+++ b/src/quickfix.c
@@ -3448,11 +3448,11 @@ ex_make(exarg_T *eap)
     }
     if (au_name != NULL)
     {
-   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf);
-   if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf))
  # ifdef FEAT_EVAL
-   if (did_throw || force_abort)
-       return;
-       if (aborting())
-       return;
  # endif
     }
  #endif
  @@ -3972,10 +3972,12 @@ ex_vimgrep(exarg_T *eap)
   }
   if (au_name != NULL)
   {
-   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf);
-   if (did_throw || force_abort)
-       return;
-   if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf))
  +#ifdef FEAT_EVAL
-       if (aborting())
-       return;
  +#endif
   }
  #endif

@@ -4877,11 +4879,11 @@ ex_cbuffer(exarg_T *eap)
     }
     if (au_name != NULL)
     {
-   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf);
-   if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf))
  # ifdef FEAT_EVAL
-   if (did_throw || force_abort)
-       return;
-       if (aborting())
-       return;
  # endif
     }
  #endif
  @@ -4968,11 +4970,11 @@ ex_cexpr(exarg_T *eap)
   }
   if (au_name != NULL)
   {
-   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf);
-   if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf))
  # ifdef FEAT_EVAL
-   if (did_throw || force_abort)
-       return;
-       if (aborting())
-       return;
  # endif
     }
  #endif
  @@ -5044,10 +5046,12 @@ ex_helpgrep(exarg_T *eap)
   }
   if (au_name != NULL)
   {
-   apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf);
-   if (did_throw || force_abort)
-       return;
-   if (au_name != NULL && apply_autocmds(EVENT_QUICKFIXCMDPRE, au_name,
-                          curbuf->b_fname, TRUE, curbuf))
  +#ifdef FEAT_EVAL
-       if (aborting())
-       return;
  +#endif
   }
  #endif

diff --git a/src/testdir/test_quickfix.vim b/src/testdir/test_quickfix.vim
index 022c12a..118a015 100644
--- a/src/testdir/test_quickfix.vim
+++ b/src/testdir/test_quickfix.vim
@@ -1631,3 +1631,20 @@ function Test_Autocmd()
          \ 'postcaddbuffer']
   call assert_equal(l, g:acmds)
 endfunction
+
+function! Test_Autocmd_Exception()
-  set efm=%m
-  lgetexpr '?'
  +
-  try
-    call DoesNotExit()
-  catch
-    lgetexpr '1'
-  finally
-    lgetexpr '1'
-  endtry
  +
-  call assert_equal('1', getloclist(0)[0].text)
  +
-  set efm&vim
  +endfunction

**Comment by blueyed** [comment]
Looks good to me, thanks!

**Comment by blueyed** [comment]
cscope is not fixed though (https://github.com/vim/vim/blob/04c4ce650f9e533cd35b2aa6803f4d354d3ec7aa/src/if_cscope.c#L1190-L1191).  Should be done in the same way, no?

**Comment by brammool** [comment]
Daniel Hahler wrote:

> cscope is not fixed though
> (https://github.com/vim/vim/blob/04c4ce650f9e533cd35b2aa6803f4d354d3ec7aa/src/if_cscope.c#L1190-L1191).
> Should be done in the same way, no?

Yes, we can do that one too.

## 

Every engineer dreams about saving the universe and having sex with aliens.
This is much more glamorous than the real life of an engineer, which consists
of hiding from the universe and having sex without the participation of other
life forms.                     (Scott Adams - The Dilbert principle)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by blueyed** [comment]
Fixed in 8.0.0068, thanks!

Do you have an idea to work around this for older Vim versions during the Neomake tests?
